examples/learned_models/pendulum.jl

```diff
@@ -1,141 +1,177 @@
-                        doplot=false, inds=1:RD.state_dim(model0))
-    times = range(0, tf, step=dt)
-    N = length(times) 
-    m = RD.control_dim(model)
-    @assert m == RD.control_dim(model0)
-    z0 = expandstate(model, x0) 
-    Z = [copy(z0) for k = 1:N]
-    X = [copy(x0) for k = 1:N]
-    for k = 1:N-1
-        RD.discrete_dynamics!(sig, model, Z[k+1], Z[k], U[k], times[k], dt)
-        RD.discrete_dynamics!(sig, model0, X[k+1], X[k], U[k], times[k], dt)
-    end
-    X_bl = map(z->model.g * z, Z)
-    X_bl, X
-    if doplot
-        X_mat = reduce(hcat, X_bl)
-        X0_mat = reduce(hcat, X)
-        p = plot(times, X0_mat[inds,:]', label="original", c=[1 2])
-        plot!(p, times, X_mat[inds,:]', label="bilinear", c=[1 2], s=:dash)
-        display(p)
-    end
-    sse = norm(X_bl - X)^2
-    println("SSE: $sse")
-    X_bl, X
+function Base.rand(params::PendulumParams) 
+    x0 = rand(params.x0) 
+    R = 1.0 
+    Q = rand(params.Qfratio)
+    Qf = 10^(rand(params.Qfratio)) * Q
+    u_bnd = rand(params.u_bnd)
+    tf_raw = rand(params.tf)
+    N = round(Int, tf_raw / params.dt) + 1
+    tf = params.dt * (N - 1)
+    (x0=x0, Qv=Q, Rv=R, Qfv=Qf, u_bnd=u_bnd, tf=tf)
 end
 
-function create_data(model::RD.DiscreteDynamics, ctrl::AbstractController, 
-                              initial_conditions, tf, dt; sig=RD.InPlace())
-    num_traj = length(initial_conditions)
-    N = round(Int, tf/dt) + 1
-    X_sim = Matrix{Vector{Float64}}(undef, N, num_traj)
-    U_sim = Matrix{Vector{Float64}}(undef, N-1, num_traj)
-    for i = 1:num_traj
-        X,U = simulatewithcontroller(sig, model, ctrl, initial_conditions[i], tf, dt)
-        X_sim[:,i] = X
-        U_sim[:,i] = U
-    end
-    X_sim, U_sim
-end
+## Generate ALTRO data
+model = RobotZoo.Pendulum()
+dmodel = RD.DiscretizedDynamics{RD.RK4}(model)
+num_traj = 100
+tf = 4.0
+dt = 0.05
 
+params_sampler = PendulumParams(tf=[tf-eps(tf),tf+eps(tf)])  # restrict it to a set horizon, for now
+opts = Altro.SolverOptions(show_summary=false)
+ctrl = ALTROController(genpendulumproblem, params_sampler, opts=opts)
+
+x0_sampler = Product([Uniform(-pi/4,pi/4), Normal(-2, 2)])
+initial_conditions = tovecs(rand(x0_sampler, num_traj), length(x0_sampler))
+X_train, U_train = create_data(dmodel, ctrl, initial_conditions, tf, dt)
+
+num_traj_test = 10
+x0_sampler_test = Product([Uniform(-eps(),eps()), Normal(-eps(), eps())])
+initial_conditions_test = tovecs(rand(x0_sampler, num_traj_test), length(x0_sampler))
+params_sampler_test = PendulumParams(tf=[tf-eps(tf),tf+eps(tf)], QRratio=[1.0, 1.1], Qfratio=[2.0,3.0], u_bnd=[4.0,5.5])
+ctrl_test = ALTROController(genpendulumproblem, params_sampler_test, opts=opts)
+X_test, U_test = create_data(dmodel, ctrl_test, initial_conditions_test, tf, dt)
+jldsave(joinpath(Problems.DATADIR, "pendulum_altro_trajectories.jld2"); X_train, U_train, X_test, U_test)
 
 ## Generate training data
 model = RobotZoo.Pendulum()
```

Comment:
Everything here and below you can ignore for now, or reference in you're curious how I'm attempting to get a good model for the pendulum.